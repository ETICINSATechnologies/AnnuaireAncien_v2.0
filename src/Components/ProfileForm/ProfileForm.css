.ProfileForm {
  display: grid;
  grid-template-rows: 2fr 2fr 1fr;
  background-color: var(--backgroundMainView);

  justify-self: center;
  vertical-align: middle;
  border: 0.6vmin solid black;
  margin: 5vmin auto;
  padding: 1vw;

  width: 90%;
  height: calc(85vh - 6.5vw - 2.5vmax);
}

.ProfileForm .header_container {
  display: grid;
  grid-template-columns: 5fr 4fr 1fr;
  grid-template-rows: 1fr;
  margin: auto;
}

.ProfileForm .field_container {
  display: grid;
  width: 100%;
  grid-template-columns: repeat(2, 1fr);
  grid-template-rows: repeat(3, 1fr);
  margin: auto;
}

.ProfileForm .field_container select {
  background-color: var(--backgroundMainView);
  position: relative;
  left: 1vw;
}

.ProfileForm .field_container input,
p {
  width: 90%;
}

.ProfileForm .button_container {
  display: grid;
  width: 100%;
  grid-template-columns: repeat(2, 1fr);
  grid-template-rows: 1fr;
  margin: 0.5vmin auto;
}

.ProfileForm .button_container input {
  width: 80%;
}

.ProfileForm .needed:after {
  content: "*";
  color: red;
}

.ProfileForm input[type="text"] {
  border: 0.3vmin solid black;
  font-size: 3vmin;
  padding-left: 1vmin;
  height: 5vmin;
}

.ProfileForm .isBoard {
  background: var(--backgroundMainView);
}
.ProfileForm p,
input[type="text"] {
  background-color: var(--backgroundMainView);
  justify-self: center;
}

.ProfileForm .right {
  grid-column: 2;
}

.ProfileForm h1 {
  text-align: center;
  justify-self: center;
  vertical-align: middle;
  width: 77%;
  font-size: 2.5vw;

  color: var(--descriptionProfileForm);

  margin: auto;
}

.ProfileForm .error_info {
  color: #eb2f06;
}

@media (orientation: portrait) {
  .ProfileForm {
    height: 68vw;
  }

  .ProfileForm .info_area {
    font-size: 3.4vw;
  }
}

.ProfileForm p {
  font-size: 3.7vmin;
  margin: auto auto 0 auto;

  height: 4.7vmin;
}

.ProfileForm .profilePicture {
  display: grid;

  margin: auto auto 0 auto;
  background-color: white;

  animation-duration: 2s;
  width: auto;
  max-height: 22vmin;
}

.ProfileForm .div_button {
  display: grid;
  justify-self: center;
  margin: auto auto 0.5vmin auto;
  height: 80%;
}

.ProfileForm .input_button {
  cursor: pointer;
  text-align: center;
  vertical-align: middle;

  margin: auto;

  font-size: 1.7vw;
  border: 0.3vmin solid black;

  height: 7vmin;
}

.ProfileForm .input_button:hover {
  cursor: pointer;
  opacity: 0.8;
}

.ProfileForm .password {
  background: #dddddd;
}

.ProfileForm .update {
  color: white;
  background-color: #009e0f;
}

.ProfileForm .deleteCancelPicture {
  width: 65%;
  margin: 1vmin auto auto;
  grid-column: 3;
  vertical-align: top;
  horiz-align: right;
}

.ProfileForm .deleteCancelPicture:hover {
  cursor: pointer;
}

.ProfileForm .modal p {
  display: grid;
}

.ProfileForm .modal a {
  display: grid;
  font-size: 3vmin;
  margin: auto auto 0 auto;
  height: 4.7vmin;
  text-align: center;
}

.ProfileForm .modal .content {
  display: grid;
  grid-template-rows: 0fr 3fr repeat(7, 2fr);
  vertical-align: top;
}

.ProfileForm .modal .message {
  display: grid;
  text-align: center;
  font-size: 3vmin;
  margin: auto auto 0 auto;
  height: 4.7vmin;
}

.ProfileForm .modal .error {
  color: red;
}

.ProfileForm .modal .success {
  color: green;
}

.ProfileForm .modal .neutral {
  color: black;
}

.ProfileForm .modal .input_button {
  position: absolute;
  background-color: #009e0f;
  color: #fff;
  bottom: 1vmin;
  left: 50%;
  transform: translate(-50%, -50%);
}

.ProfileForm .modal input[type="password"] {
  display: grid;
  margin: 0 auto;
  background-color: white;
  justify-self: center;
  border: 0.3vmin solid black;
  font-size: 3vmin;
  padding-left: 1vmin;
  height: 5vmin;
}
