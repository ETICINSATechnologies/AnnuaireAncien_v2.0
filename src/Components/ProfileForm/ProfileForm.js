import React, {Component} from 'react';
import './ProfileForm.css';
import Auth from "../Auth/Auth";
import modifyIcon from "../../Images/edit.png";
import cancelIcon from "../../Images/cancel_icon.png";
import noPhotoIcon from "../../Images/no_photo.png";

class ProfileForm extends Component {
    constructor(props) {
        super(props);
        this.state = {
            modifyEnabled: false,
            departments: [],
            info: {
                firstName: '',
                lastName: '',
                email: '',
                telephone: '',
                department: {
                    id: 0,
                    label: '',
                    name: '',
                },
                company: '',
            }
        };
        this.onChange = this.onChange.bind(this);
    }


    onChange(event) {
        event.persist();
        if (this.state.info.hasOwnProperty(event.target.className)) {
            if (event.target.className === 'department') {
                this.setState(({
                    info: {
                        ...this.state.info,
                        department: {
                            ...this.state.info.department,
                            id: parseInt(event.target.value, 10)
                        }
                    }
                }));
            }
            else {
                this.setState(({
                    info: {
                        ...this.state.info,
                        [event.target.className]: event.target.value
                    }
                }));
            }
        }
    }

    modifyProfile() {
        let data;
        data = {
            ...this.state.info,
            genderId: this.props.info.gender.id,
            address: {
                ...this.state.info.address,
                countryId: this.props.info.address.country.id
            }

        };

        if(this.state.info.department===null){
            data.departmentId=null;
        } else {
            data.departmentId=this.state.info.department.id;
        }

        this.props.update(data);
    }

    modifyEnable() { //does both enable and disable
        this.props.setModify(!this.state.modifyEnabled);
    }

    componentDidUpdate(prevProps) {
        if (this.props.info !== prevProps.info || this.props.modifyEnabled !== prevProps.modifyEnabled) {
            this.setState({
                info: this.props.info,
                modifyEnabled: this.props.modifyEnabled,
            });
        }
    }

    componentDidMount() {
        fetch('api/v1/core/department', {
            headers: {
                Authorization: Auth.getToken()
            }
        })
            .then(res => res.json())
            .then((result) => {
                this.setState({departments: result});
            })
    }

    renderDepartments() {
        let departmentDropDown = this.state.departments.map((department, index) => {
            return <option key={index} value={department.id}>{department.label}</option>
        });
        return (
            <select className="department"
                    value={this.state.info.department === null || typeof(this.state.info.department) === 'undefined' ?
                        0 :
                        this.state.info.department.id}
                    onChange={this.onChange}
                    disabled={!this.state.modifyEnabled}
            >
                <option value={0}> Choisir un departement</option>
                {departmentDropDown}
            </select>
        )
    }

    render() {
        return (
            <form className="ProfileForm">
                <div className="header_container">
                    <h1>
                        {
                            this.state.modifyEnabled ?
                            "Appuyer sur la croix pour annuler" :
                            "Appuyer sur le crayon pour modifier"
                        }
                    </h1>
                    <img className="image" src={noPhotoIcon} alt="Profile"/>
                    <img className="modifier"
                         src={this.state.modifyEnabled ? cancelIcon : modifyIcon}
                         onClick={() => {
                             this.state.modifyEnabled ? this.props.resetFields() : this.modifyEnable()
                         }} alt="Modifier/Annuler"
                    />
                </div>
                <div className="field_container">
                    <p> Téléphone </p>
                    <p className="right needed"> Nom </p>
                    <input disabled={!this.state.modifyEnabled} type="text" className="telephone"
                           value={this.state.info.telephone} onChange={this.onChange}/>
                    <input disabled={!this.state.modifyEnabled} type="text" className="lastName"
                           value={this.state.info.lastName} onChange={this.onChange}/>
                    <p> Département </p>
                    <p className="right p_info needed"> Prénom </p>
                    {this.renderDepartments()}
                    <input disabled={!this.state.modifyEnabled} type="text" className="firstName"
                           value={this.state.info.firstName} onChange={this.onChange}/>
                    <p className="needed"> Adresse mail </p>
                    <p> Travaille chez </p>
                    <input disabled={!this.state.modifyEnabled} type="text" className="email"
                           value={this.state.info.email} onChange={this.onChange}/>
                    <input disabled={!this.state.modifyEnabled} type="text" className="company"
                           value={this.state.info.company} onChange={this.onChange}/>
                </div>
                <div className="button_container">
                    {
                        this.state.modifyEnabled ?
                            <React.Fragment>
                                <Button className="input_button password" value="Changer mot de passe"/>
                                <Button className="input_button update" value="Sauvegarder"
                                        onClick={() => this.modifyProfile()}/>
                            </React.Fragment> :
                            null
                    }
                </div>

            </form>
        );
    }
}

function Button(props) {
    return (
        <input className={props.className} type="button" value={props.value} onClick={props.onClick}/>
    );
}

export default ProfileForm;
