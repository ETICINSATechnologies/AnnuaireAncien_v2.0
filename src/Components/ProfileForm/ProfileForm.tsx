import React, {Component} from 'react';
import './ProfileForm.css';
import Auth from "../Auth/Auth";
import {Mdp} from "../../Model/Mdp";
import {Member} from "../../Model/Member";
import {Link} from "react-router-dom";
import Modal from "../Modal/Modal";
import DropDown from "../DropDown/DropDown";
import ProfilePicture from "../ProfilePicture/ProfilePicture";

let modifyIcon = require("../../Images/edit.png");
let cancelIcon = require("../../Images/cancel_icon.png");
let manIcon = require("../../Images/default_man.svg");
let womanIcon = require("../../Images/default_woman.svg");


interface ProfileFormProps {
    member: Member
    modifyEnabled: boolean
    update: boolean
    updateSucceed: boolean

    resetFields(): void

    modifyMember(member: Member): void

    updateMember(): void

    updateMemberPassword(pass: string): void

    enableModification(): void
}

interface ProfileFormState {
    idMapping: object
    show: boolean
    editImageShow: boolean
    mdp: Mdp
    downloadedImg: string
}

const genders = [
    {id: 1, label: 'Autre'}, {id: 2, label: 'M'}, {id: 3, label: 'F'}
];


export class ProfileForm extends Component<ProfileFormProps, ProfileFormState> {
    state = {
        idMapping: [],
        show: false,
        editImageShow: false,
        mdp: {
            mdpancien: '',
            mdpnouveau: '',
            mdpnouveau2: '',
            mdpstate: 'Compléter les champs et appuyer sur valider',
            mdpstatetype: 'neutral',
        },
        downloadedImg: ''
    };

    getMemberImage = () => {
        fetch('member/me/image', {
            headers: {
                'Authorization': Auth.getToken(),
                'Content-Type': 'application/json'
            }
        })
            .then(res => {
                if (res.status === 200) {
                    res.blob()
                        .then(image => {
                            this.setState({
                                downloadedImg: URL.createObjectURL(image)
                            })
                        })
                } else {
                    this.setState({downloadedImg: ''});
                }
            })
    };

    componentDidMount() {
        this.getMemberImage()
    }

    displayImage = () => {
        let member = this.props.member;
        if (this.state.downloadedImg)
            return this.state.downloadedImg;

        if (member) {
            if (member.gender === 'F')
                return womanIcon;
            else if (member.gender === 'M')
                return manIcon;
        }
    };

    onChange = (event: React.ChangeEvent) => {
        event.persist();
        let property = event.target.className;
        let value = (event.target as any).value;
        let member = new Member(this.props.member);

        if (member.hasOwnProperty(property)) {
            property === 'gender' ?
                member[property] = this.getGenderLabel(value) : member[property] = value;
        }

        this.props.modifyMember(member);
    };

    onChangeMdp = (event: React.ChangeEvent) => {
        event.persist();
        this.setState(({
            mdp: {
                ...this.state.mdp,
                [event.target.className]: (event.target as any).value
            }
        }));
    };

    updateMdp = () => {
        // check if new pass is at least 8 char
        if (this.state.mdp.mdpnouveau.length < 8) {
            this.setState({
                mdp: {
                    ...this.state.mdp,
                    mdpstate: 'Ce mot de passe est trop court',
                    mdpstatetype: 'error'
                }
            });
            return 0;
        }

        // check if two new passwords match
        if (this.state.mdp.mdpnouveau !== this.state.mdp.mdpnouveau2) {
            this.setState({
                mdp: {
                    ...this.state.mdp,
                    mdpstate: 'Les mots de passe ne corrrespondent pas',
                    mdpstatetype: 'error'
                }
            });
            return 0;
        }

        // make username + mdp pair
        let pair = {
            username: this.props.member.username,
            password: this.state.mdp.mdpancien
        };

        // check old pass
        fetch('login', {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify(pair)
        })
            .then(res => {
                if (res.status === 200) {
                    this.props.updateMemberPassword(this.state.mdp.mdpnouveau);
                    this.setState({
                        mdp: {
                            ...this.state.mdp,
                            mdpstate: 'Le mot de passe a bien été modifié',
                            mdpstatetype: 'success'
                        }
                    });
                }
                else {
                    this.setState({
                        mdp: {
                            ...this.state.mdp,
                            mdpstate: "L'ancien mot de passe n'est pas valide",
                            mdpstatetype: 'error'
                        }
                    });
                    return 0;
                }
            })
    };

    showModal = () => {
        this.setState({
            ...this.state,
            show: !this.state.show,
            mdp: {
                ...this.state.mdp,
                mdpstate: 'Compléter les champs et appuyer sur valider',
                mdpstatetype: 'neutral'
            }
        })
    };

    getGenderId(gender: string): number {
        switch (gender) {
            case 'Autre':
                return 1;
            case 'M':
                return 2;
            case 'F':
                return 3;
            default :
                return 0;
        }
    }

    getGenderLabel(id: string): string {
        switch (id) {
            case '1':
                return 'Autre';
            case '2':
                return 'M';
            case '3':
                return 'F';
            default:
                return 'Choisir une option';
        }
    }

    displayEditImage = (display: boolean) => {
        if (!display)
            this.getMemberImage();
        this.setState({
            editImageShow: display
        })
    };

    render() {
        return (
            <form className="ProfileForm">
                <div className="header_container">
                    {
                        this.props.update ?
                            this.props.updateSucceed ?
                                <h2>Les modifications ont bien été enregistrées</h2> :
                                <h2 className='error_info'>Une erreur est survenue lors de la sauvegarde</h2> :
                            this.props.modifyEnabled ?
                                <h2>Appuyer sur la croix pour annuler</h2> :
                                <h2>Appuyer sur le crayon pour modifier</h2>
                    }
                    <img className="profilePicture" src={this.displayImage()} alt=""
                         onClick={() => this.displayEditImage(true)}/>
                    <ProfilePicture show={this.state.editImageShow} onClose={this.displayEditImage}/>
                    <img className="deleteCancelPicture"
                         src={this.props.modifyEnabled ? cancelIcon : modifyIcon}
                         onClick={() => this.props.enableModification()} alt="Modifier/Annuler"/>
                    <p className="gender"> Genre </p>
                    <DropDown className='gender' onChange={this.onChange} modifyEnabled={this.props.modifyEnabled}
                              options={genders} currentOption={this.getGenderId(this.props.member.gender)}/>
                </div>
                <div className="field_container">
                    <p className="needed"> Prénom </p>
                    <p className="needed"> Nom </p>

                    <input disabled={!this.props.modifyEnabled} type="text" className="firstName"
                           value={this.props.member.firstName} onChange={this.onChange}/>
                    <input disabled={!this.props.modifyEnabled} type="text" className="lastName"
                           value={this.props.member.lastName} onChange={this.onChange}/>
                    <p> Téléphone </p>
                    <p className="needed"> Adresse mail </p>
                    <input disabled={!this.props.modifyEnabled} type="text" className="telephone"
                           value={this.props.member.telephone} onChange={this.onChange}/>
                    <input disabled={!this.props.modifyEnabled} type="text" className="email"
                           value={this.props.member.email} onChange={this.onChange}/>
                    <p> Année de sortie </p>
                    <p> Date de naissance </p>
                    <input disabled={!this.props.modifyEnabled} type="text" className="gradeYear"
                           value={this.props.member.gradeYear} onChange={this.onChange}/>
                    <input type="date" disabled={!this.props.modifyEnabled} className="birthday"
                           value={this.props.member.birthday} onChange={this.onChange}/>
                    <p> LinkedIn </p>
                    <p> Facebook </p>
                    <input disabled={!this.props.modifyEnabled} type="text" className="linkedin"
                           value={this.props.member.linkedin? this.props.member.linkedin : ''} 
                           onChange={this.onChange}/>
                    <input disabled={!this.props.modifyEnabled} type="text" className="facebook"
                           value={this.props.member.facebook? this.props.member.facebook : ''} 
                           onChange={this.onChange}/>
                </div>
                <div className="button_container">
                    {
                        !this.props.modifyEnabled ? null :
                            <React.Fragment>
                                <input type="button" className="input_button password" value="Changer mot de passe"
                                       onClick={() => this.showModal()}/>
                                <input type="button" className="input_button update" value="Sauvegarder"
                                       onClick={() => this.props.updateMember()}/>
                            </React.Fragment>
                    }
                </div>
                <div className="modal">
                    <Modal show={this.state.show} onClose={this.showModal}>
                        <div className="content">
                            <p className={"message " + this.state.mdp.mdpstatetype}> {this.state.mdp.mdpstate} </p>
                            <p> Ancien mot de passe </p>
                            <input type="password" name="password" className="mdpancien"
                                   onChange={this.onChangeMdp.bind(this)}/>
                            <p> Nouveau mot de passe </p>
                            <input type="password" name="password" className="mdpnouveau"
                                   onChange={this.onChangeMdp.bind(this)}/>
                            <p> Nouveau mot de passe </p>
                            <input type="password" name="password" className="mdpnouveau2"
                                   onChange={this.onChangeMdp.bind(this)}/>
                            <Link to="/recovery"> Mot de passe oublié ?</Link>
                            <input type="button" className="input_button" value="Valider"
                                   onClick={this.updateMdp.bind(this)}/>
                        </div>
                    </Modal>
                </div>
            </form>
        );
    }

}

export default ProfileForm;