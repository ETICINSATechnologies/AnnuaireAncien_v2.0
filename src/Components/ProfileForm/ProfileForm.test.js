import React from 'react';
import ProfileForm from './ProfileForm';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';


Enzyme.configure({adapter: new Adapter()});

describe( 'ProfileForm component', () => {
    it('renders without crashing', () => {
            const tree = renderer.create(<ProfileForm />).toJSON();
            expect(tree).toMatchSnapshot();
        }
    );

    it('ProfileForm unit test', () => {
        expect(Enzyme.shallow(<ProfileForm/>)
            .find('.ProfileForm')
            .length
        ).toBe(1)
    });

});