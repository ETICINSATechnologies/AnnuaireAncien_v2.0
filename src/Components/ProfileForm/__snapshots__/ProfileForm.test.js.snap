// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfileForm component renders without crashing 1`] = `
<form
  className="ProfileForm"
>
  <div
    className="info_area"
  >
    Appuyer sur Sauvegarder pour enregistrer les modifications
  </div>
  <div
    className="image"
  >
     
  </div>
  <p>
     Téléphone 
  </p>
  <p
    className="right needed"
  >
     Nom 
  </p>
  <input
    className="telephone"
    onChange={[Function]}
    type="text"
    value=""
  />
  <input
    className="lastName"
    onChange={[Function]}
    type="text"
    value=""
  />
  <p>
     Département 
  </p>
  <p
    className="right p_info needed"
  >
     Prénom 
  </p>
  <input
    className="department"
    onChange={[Function]}
    type="text"
    value=""
  />
  <input
    className="firstName"
    onChange={[Function]}
    type="text"
    value=""
  />
  <p
    className="needed"
  >
     Adresse mail 
  </p>
  <p>
     Travaille chez 
  </p>
  <input
    className="email"
    onChange={[Function]}
    type="text"
    value=""
  />
  <input
    className="company"
    onChange={[Function]}
    type="text"
    value=""
  />
  <div
    className="div_button"
  >
    <input
      className="input_button password"
      type="button"
      value="Changer mot de passe"
    />
  </div>
  <div
    className="div_button"
  >
    <input
      className="input_button update"
      type="button"
      value="Sauvegarder"
    />
  </div>
</form>
`;
